### PR TITLE
Payfi vaults

### PR DESCRIPTION
**Pull request by lavacakechef**
- [x] Add deposit cap for teller to limit total deposit values
- [x] Add interest rate accrual for exchange rate (NAV) and separate management fee inside Accountant

### CODE REVIEW

**Comment by VarinderSS1113** [comment]
Hi @jingyi2811 

Could you please help to review the contract changes please, thanks!

**Comment by jingyi2811** [comment]
I believe the core code changes are in AccountantWithRateProviders.sol, so I took a closer look at the file.

**Comment by jingyi2811** [comment]
I'm not sure if the design intention is for interest to continue accruing when the contract is paused.

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 1]
```diff
@@ -283,6 +322,54 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
+        lendingInfo.lendingRate = _lendingRate;
+        lendingInfo.lastAccrualTime = block.timestamp;
+        emit LendingRateUpdated(_lendingRate, block.timestamp);
+    }
+
+    /**
+     * @notice Set protocol fee rate (cheap - no checkpoint)
+     */
+    function setProtocolFeeRate(uint256 _protocolFeeRate) external requiresAuth {
+        lendingInfo.protocolFeeRate = _protocolFeeRate;
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: Like in updateExchangeRate() and setLendingRate(), I believe setProtocolFeeRate() should accrue protocol fees before setting the new rate to ensure accounting is consistent.

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 2]
```diff
@@ -283,6 +322,54 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
+        emit ProtocolFeeRateUpdated(_protocolFeeRate, block.timestamp);
+    }
+
+    /**
+     * @notice Set maximum lending rate
+     * @dev Callable by OWNER_ROLE
+     */
+    function setMaxLendingRate(uint256 _maxLendingRate) external requiresAuth {
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol:320]: I think we can remove this function. Both setLendingRate and setMaxLendingRate use requiresAuth. That means the max cap might not be necessary, since trusted users are expected to know the right values when calling setLendingRate. It's like letting the same person set a restriction value to prevent themselves from making a mistake which is unnecessary

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 3]
```diff
@@ -294,6 +381,16 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
 
         AccountantState storage state = accountantState;
         if (state.isPaused) revert AccountantWithRateProviders__Paused();
+
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: I think we can create a function to avoid duplicating code:

```
if (vault.totalSupply() > 0 && lendingInfo.lendingRate > 0) {
    uint256 timeElapsed = block.timestamp - lendingInfo.lastAccrualTime;
    uint256 totalDeposits = vault.totalSupply().mulDivDown(accountantState.exchangeRate, ONE_SHARE);

    uint256 interestAccrued = totalDeposits.mulDivDown(
        lendingInfo.lendingRate * timeElapsed,
        SECONDS_PER_YEAR * BASIS_POINTS
    );

    uint256 protocolFees = totalDeposits.mulDivDown(
        lendingInfo.protocolFeeRate * timeElapsed,
        SECONDS_PER_YEAR * BASIS_POINTS
    );

    accountantState.feesOwedInBase += uint128(protocolFees);
    lendingInfo.lastAccrualTime = block.timestamp;
}
```

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 4]
```diff
@@ -44,6 +44,17 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
         uint16 managementFee;
     }
 
+    /**
+     * @notice Lending specific state
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol:47]: The comment seems wrong because it doesn't mention lendingRate and protocolFeeRate.

**Comment by lavacakechef** [comment]
> I'm not sure if the design intention is for interest to continue accruing when the contract is paused.

@jingyi2811 this is design choice that I discussed with @VarinderClearpool 
So I think we can keep as is, but document this behavior

**Comment by lavacakechef** [comment]
<html><head></head><body><h1>Keyring Integration Implementation Plan</h1>
<h2>Overview</h2>
<p>Integrate dual-mode access control (Keyring KYC / Manual Whitelist) into TellerWithMultiAssetSupport for mint/redeem operations.</p>
<h2>Phase 1: Core Infrastructure (Day 1)</h2>
<p><strong>Time: 6-8 hours</strong></p>
<h3>1.1 Add Keyring Interface (1 hour)</h3>
<pre><code class="language-solidity">interface IKeyring {
    function checkCredential(uint256 policyId, address entity) external view returns (bool);
}
</code></pre>
<h3>1.2 Update TellerWithMultiAssetSupport State (2 hours)</h3>
<pre><code class="language-solidity">enum AccessControlMode {
    DISABLED,
    KEYRING_KYC,
    MANUAL_WHITELIST
}

AccessControlMode public accessControlMode;
IKeyring public keyringContract;
uint256 public keyringPolicyId;
mapping(address =&gt; bool) public manualWhitelist;
mapping(address =&gt; bool) public contractWhitelist; // For AMMs/contracts
</code></pre>
<h3>1.3 Implement Access Control Modifier (2 hours)</h3>
<pre><code class="language-solidity">modifier checkAccess(address entity) {
    if (accessControlMode == AccessControlMode.KEYRING_KYC) {
        // Check Keyring unless whitelisted contract
    } else if (accessControlMode == AccessControlMode.MANUAL_WHITELIST) {
        // Check manual whitelist
    }
    _;
}
</code></pre>
<h3>1.4 Admin Functions (1-2 hours)</h3>
<ul>
<li><code>setAccessControlMode()</code></li>
<li><code>setKeyringConfig()</code></li>
<li><code>updateManualWhitelist()</code></li>
<li><code>updateContractWhitelist()</code></li>
</ul>
<h2>Phase 2: Integration Points (Day 2)</h2>
<p><strong>Time: 6-8 hours</strong></p>
<h3>2.1 Apply to Deposit Functions (3 hours)</h3>
<ul>
<li><code>deposit()</code> - Check <code>msg.sender</code></li>
<li><code>depositWithPermit()</code> - Check <code>msg.sender</code></li>
<li><code>bulkDeposit()</code> - Check <code>to</code> address parameter</li>
</ul>
<h3>2.2 Apply to Withdrawal Functions (2 hours)</h3>
<ul>
<li><code>bulkWithdraw()</code> - Check <code>msg.sender</code></li>
</ul>
<h3>2.3 Edge Case Handling (3 hours)</h3>
<ul>
<li>Zero address checks</li>
<li>Mode transition logic</li>
<li>Emergency bypass considerations</li>
<li>Gas optimization for repeated checks</li>
</ul>
<h2>Phase 3: Testing Suite (Day 3)</h2>
<p><strong>Time: 8-10 hours</strong></p>
<h3>3.1 Unit Tests (4 hours)</h3>
<pre><code class="language-solidity">// Test scenarios:
- testKeyringModeAllowsValidCredentials()
- testKeyringModeBlocksInvalidCredentials()
- testManualWhitelistMode()
- testDisabledMode()
- testContractWhitelistInBothModes()
- testModeTransitions()
- testBulkDepositToParameter()
</code></pre>
<h3>3.2 Integration Tests (3 hours)</h3>
<ul>
<li>Fork mainnet testing with real Keyring contract</li>
<li>Test with Policy ID 7 on Sepolia</li>
<li>Multi-user scenario testing</li>
<li>Gas usage benchmarking</li>
</ul>
<h3>3.3 Security Tests (3 hours)</h3>
<ul>
<li>Access control bypass attempts</li>
<li>Reentrancy with access checks</li>
<li>Mode switching edge cases</li>
<li>Whitelist manipulation tests</li>
</ul>
<h2>Phase 4: Deployment &amp; Documentation (Day 4)</h2>
<p><strong>Time: 4-6 hours</strong></p>
<h3>4.1 Deployment Scripts (2 hours)</h3>
<pre><code class="language-javascript">// Deploy script updates for each network
const KEYRING_ADDRESSES = {
    mainnet: "0x51d67d4f1b46e69b7bfd48d9f80aba74817eba9a",
    arbitrum: "0xdf5ebf1bc5782ba01dfefd2f07556ee017b4ce65",
    base: "0x968aa1818af69e6ca8f000b8fb7a981c227e3684",
    // ... other networks
};
</code></pre>
<h3>4.2 Configuration Documentation (2 hours)</h3>
<ul>
<li>Admin operation guide</li>
<li>Mode selection criteria</li>
<li>Whitelist management procedures</li>
<li>Emergency response procedures</li>
</ul>
<h3>4.3 Deployment Verification (2 hours)</h3>
<ul>
<li>Cross-chain deployment verification</li>
<li>Initial configuration setup</li>
<li>Smoke tests on each network</li>
</ul>
<h2>Total Estimation</h2>

Phase | Duration | Hours
-- | -- | --
Phase 1: Core Infrastructure | 1 day | 6-8 hours
Phase 2: Integration Points | 1 day | 6-8 hours
Phase 3: Testing Suite | 1 day | 8-10 hours
Phase 4: Deployment | 1 day | 4-6 hours
Total | 4 days | 24-32 hours


<h2>Risk Factors &amp; Mitigation</h2>
<ol>
<li><strong>Cross-chain Keyring availability</strong>: Test early with provided addresses</li>
<li><strong>Gas costs</strong>: Implement caching if needed</li>
<li><strong>Mode transition timing</strong>: Clear procedures for switching modes</li>
<li><strong>Contract whitelist management</strong>: Comprehensive initial list of DEXs/protocols</li>
</ol>
<h2>Success Criteria</h2>
<p>✅ All deposit/withdrawal functions check access control<br>
✅ Both Keyring and manual modes work correctly<br>
✅ Transfers remain unrestricted<br>
✅ Per-vault configuration is operational<br>
✅ Gas increase is &lt; 15% for operations<br>
✅ All tests pass on mainnet fork<br>
✅ Successfully deployed on all target chains</p></body></html>

**Comment by lavacakechef** [comment]
## **Summary of Changes**

### **AccountantWithRateProviders.sol**

**Added Lending & Fee System:**
* Auto-compounding interest accrual on vault deposits (`lendingRate`)
* Single management fee paid by borrowers (`_managementFee` in `accountantState`)
* Virtual accrual with real-time calculation via `calculateExchangeRateWithInterest()`
* Helper functions `_checkpointManagementFees()` and `_checkpointInterestAndFees()` to handle state updates
* Management fee checkpoint on rate changes and claims

**Key Design:**
* Interest increases exchange rate automatically (benefits depositors)
* Management fees stored separately in `_feesOwedInBase` 
* Borrowers pay: `lendingRate + _managementFee`
* Protocol earns: `_managementFee` only
* Works with `lendingRate = 0` as traditional vault (management fee only)
* Admin can still manually update exchange rate via `updateExchangeRate()` which checkpoints fees first

**New Admin Functions:**
* `setLendingRate()` - Configure depositor APY
* `setManagementFeeRate()` - Configure protocol fee
* `setMaxLendingRate()` - Set rate ceiling
* `getBorrowerRate()` - View total borrower cost
* `previewFeesOwed()` - Real-time fee calculation

### **TellerWithMultiAssetSupport.sol**

**Added Access Control System:**
* Three modes: `DISABLED`, `KEYRING_KYC`, `MANUAL_WHITELIST`
* Keyring integration for automated KYC checks (Policy ID configurable)
* Manual whitelist for direct address control
* Contract whitelist for AMMs/protocols (works in all modes)
* Applied to `deposit()`, `depositWithPermit()`, `bulkDeposit()`, `bulkWithdraw()`

**Added Deposit Cap:**
* Global cap in base asset value
* Enforced on every deposit by calculating share value
* Configurable via `setDepositCap()`

**Technical Updates:**
* Both contracts upgraded to Solidity 0.8.22
* Added `checkAccess` modifier for all mint/redeem operations
* Maintained backward compatibility for existing integrations

File: src/base/Roles/TellerWithMultiAssetSupport.sol [hunk 5]
```diff
@@ -304,6 +414,7 @@ contract TellerWithMultiAssetSupport is Auth, BeforeTransferHook, ReentrancyGuar
         external
         requiresAuth
         nonReentrant
+        checkAccess(to)
```
Comment (jingyi2811) [inline src/base/Roles/TellerWithMultiAssetSupport.sol]: This is the only one obvious bug

checkAccess(to) checks if the recipient is whitelisted, but should check if the caller is authorized to deposit.

KIndly refer to https://github.com/Ozean-L2/nucleus-boring-vault/blob/feature/payfi_vaults/src/base/Roles/TellerWithMultiAssetSupport.sol#L430-L448

**Comment by jingyi2811** [comment]
Below are other low or informational findings

**Comment by jingyi2811** [comment]
All function input parameters should start with the `_` prefix, as in your other code.

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 6]
```diff
@@ -371,7 +478,94 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
+     * @return Total borrower rate in basis points
+     */
+    function getBorrowerRate() public view returns (uint256) {
+        return lendingInfo.lendingRate + lendingInfo.protocolFeeRate;
+    }
+
+    /**
+     * @notice Get asset's base rate without vault exchange rate
+     */
+    function getAssetToBaseRate(ERC20 asset) public view returns (uint256 rate) {
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: You pass in the asset as a parameter but use the global decimal. This function would be more flexible if it retrieved the decimal directly from the asset

File: src/base/Roles/TellerWithMultiAssetSupport.sol (hunk 5 above)
Comment (lavacakechef) [reply src/base/Roles/TellerWithMultiAssetSupport.sol]: this is intended behaviour as:
```
1. deposit/depositWithPermit:
msg.sender deposits AND receives shares → check msg.sender
These are public functions (despite having requiresAuth, they're set as public capabilities)

2. bulkDeposit:
msg.sender deposits BUT to receives shares → check to
Only ADMIN/SOLVER roles can call this

3. bulkWithdraw:
msg.sender burns shares → check msg.sender
Only ADMIN/SOLVER roles can call this
Assets go to to (doesn't need KYC)

Idea:
Minting shares: Check the recipient (KYC/whitelist required)
Burning shares: Check the holder (KYC/whitelist required)
Transferring assets: No checks needed
```

File: src/base/Roles/AccountantWithRateProviders.sol (hunk 6 above)
Comment (lavacakechef) [reply src/base/Roles/AccountantWithRateProviders.sol]: make sense. I remove the function as I think it doesnt serve any purpose for now

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 7]
```diff
@@ -36,14 +35,25 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
         uint128 feesOwedInBase;
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: Other parameters should start with _ too

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 8]
```diff
@@ -117,9 +137,9 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
         address payoutAddress,
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: Same

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 9]
```diff
@@ -312,21 +387,50 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
+            uint256 timeElapsed = block.timestamp - lendingInfo.lastAccrualTime;
+
+            // Calculate interest on total deposits
+            uint256 totalDeposits = vault.totalSupply().mulDivDown(accountantState.exchangeRate, ONE_SHARE);
+            interestAccrued =
+                totalDeposits.mulDivDown(lendingInfo.lendingRate * timeElapsed, SECONDS_PER_YEAR * BASIS_POINTS);
+
+            // Update rate (no fee deduction for lending model)
+            uint256 totalSupply = vault.totalSupply();
+            if (totalSupply > 0) {
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: ```if (totalSupply > 0) {``` can be removed

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 10]
```diff
@@ -366,22 +474,93 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
+
     /**
      * @notice Used to change the decimals of precision used for an amount.
      */
-    function changeDecimals(uint256 amount, uint8 fromDecimals, uint8 toDecimals) internal pure returns (uint256) {
-        if (fromDecimals == toDecimals) {
-            return amount;
-        } else if (fromDecimals < toDecimals) {
-            return amount * 10 ** (toDecimals - fromDecimals);
+    function changeDecimals(uint256 _amount, uint8 _fromDecimals, uint8 _toDecimals) internal pure returns (uint256) {
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: Internal function should start with _ like https://github.com/Ozean-L2/nucleus-boring-vault/pull/2/files#diff-b40fbd70955504478372516aed7abfb854b66064a9ced21b8861558e192d0883R531

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 11]
```diff
@@ -312,21 +387,50 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
-        // Zero out fees owed.
+
+        // Zero out fees owed
         state.feesOwedInBase = 0;
-        // Transfer fee asset to payout address.
-        feeAsset.safeTransferFrom(msg.sender, state.payoutAddress, feesOwedInFeeAsset);
 
-        emit FeesClaimed(address(feeAsset), feesOwedInFeeAsset);
+        // Transfer fee asset to payout address
+        _feeAsset.safeTransferFrom(msg.sender, state.payoutAddress, feesOwedInFeeAsset);
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol:395]: I don't see claimFees function from BoringVault.sol

Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]: AccountantWithRateProviders.sol looks fine to me so far.

One more thing—did you try approving from BoringVault first and then calling claimFee so that safeTransferFrom can work?

Comment (lavacakechef) [reply src/base/Roles/AccountantWithRateProviders.sol]: what do you mean? approve token spending or claim approval?

Comment (jingyi2811) [reply src/base/Roles/AccountantWithRateProviders.sol]: claimFees can only be called by the vault, right? Have you tried calling any function from the Vault that then triggers this function?

**Comment by jingyi2811** [comment]
So we can update _lastAccrualTime without accruing interest?

File: src/atomic-queue/AtomicQueue.sol [hunk 12]
```diff
@@ -202,47 +187,76 @@ contract AtomicQueue is ReentrancyGuard {
...
+                want.safeTransferFrom(solver, users[i], amountOut);
 
                 emit AtomicRequestFulfilled(
-                    users[i], address(offer), address(want), request.offerAmount, assetsToUser, block.timestamp
+                    users[i], address(offer), address(want), request.offerAmount, amountOut, block.timestamp
                 );
 
-                // Set shares to withdraw to 0.
                 request.offerAmount = 0;
                 request.inSolve = false;
```
Comment (jingyi2811) [inline src/atomic-queue/AtomicQueue.sol:262]: Can you take a look at the code here should we add `request.deadline=0`

Comment (jingyi2811) [reply src/atomic-queue/AtomicQueue.sol:262]: It seems like this suggestion is more of a nice-to-have improvement rather than a critical fix. It doesn’t cause significant harm but mainly enhances readability.

File: src/atomic-queue/AtomicQueue.sol [hunk 13]
```diff
@@ -135,47 +139,28 @@ contract AtomicQueue is ReentrancyGuard {
...
-     *         To determine a safe atomicPrice, share.previewRedeem should be used to get
-     *         a good share price, then the user can lower it from there to make their request fill faster.
      * @param offer the ERC20 token the user is offering in exchange for the want
      * @param want the ERC20 token the user wants in exchange for offer
-     * @param userRequest the users request
+     * @param deadline unix timestamp for when request is no longer valid
+     * @param offerAmount the amount of offer asset to exchange
      */
-    function updateAtomicRequest(ERC20 offer, ERC20 want, AtomicRequest calldata userRequest) external nonReentrant {
+    function updateAtomicRequest(ERC20 offer, ERC20 want, uint64 deadline, uint96 offerAmount) external nonReentrant {
```
Comment (jingyi2811) [inline src/atomic-queue/AtomicQueue.sol:157]: isAtomicRequestValid should logically be applied  to both updateAtomicRequest and solve function

For solve:

Current behavior:
_prepareSolve already reverts if:
inSolve == true
deadline < block.timestamp
offerAmount == 0

Missing checks:
balanceOf(user) >= offerAmount
allowance(user, this) >= offerAmount

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 14]
```diff
@@ -241,116 +251,184 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
-        state.feesOwedInBase = 0;
-        // Transfer fee asset to payout address.
-        feeAsset.safeTransferFrom(msg.sender, state.payoutAddress, feesOwedInFeeAsset);
 
-        emit FeesClaimed(address(feeAsset), feesOwedInFeeAsset);
+        // Zero out fees owed
+        state._feesOwedInBase = 0;
+
+        // Transfer fee asset to payout address
+        _feeAsset.safeTransferFrom(msg.sender, state._payoutAddress, feesOwedInFeeAsset);
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol:362]: Final observation: 

Currently, safeTransferFrom moves the vault token to the payout address, which requires two steps—approve and then safeTransferFrom. 

Instead, we could add a function in the vault that performs this transfer (No more safeTransferFrom use safeTransfer) directly, allowing AtomicQueue to call it in a single step

File: src/atomic-queue/AtomicQueue.sol (hunk 13 above)
Comment (lavacakechef) [reply src/atomic-queue/AtomicQueue.sol:157]: **updateAtomicRequest**:
It is possible for a withdraw request to return false from this function, but using the request in updateAtomicRequest will succeed

**_prepareSolve**:
safeTransfer can check for allowance and balance, isAtomicRequestValid is an external helper function and can remain so to save gas, also it takes calldata instead of storage

**Comment by jingyi2811** [comment]
Hi Spearbit / Cantina,

The files are

AccountantWithRateProviders.sol
TellerWithMultiAssetSupport.sol
AtomicQueue.sol

You can start auditing.

Thanks!

File: src/base/Roles/AccountantWithRateProviders.sol [hunk 15]
```diff
@@ -241,116 +251,184 @@ contract AccountantWithRateProviders is Auth, IRateProvider {
...
+            // Calculate interest on total deposits
+            uint256 totalDeposits = vault.totalSupply().mulDivDown(accountantState._exchangeRate, ONE_SHARE);
+            interestAccrued =
+                totalDeposits.mulDivDown(lendingInfo._lendingRate * timeElapsed, SECONDS_PER_YEAR * BASIS_POINTS);
+
+            // Update rate (no fee deduction for lending model)
+            uint256 totalSupply = vault.totalSupply();
+
+            uint256 rateIncrease = interestAccrued.mulDivDown(ONE_SHARE, totalSupply);
+            newRate += uint96(rateIncrease);
```
Comment (jingyi2811) [inline src/base/Roles/AccountantWithRateProviders.sol]:
```
newRate = uint96(
    uint256(newRate).mulDivDown(
        BASIS_POINTS + (lendingInfo._lendingRate * timeElapsed / SECONDS_PER_YEAR),
        BASIS_POINTS
    )
);
```
